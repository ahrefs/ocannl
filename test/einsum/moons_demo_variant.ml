open Base
open Ocannl
module IDX = Train.IDX
module TDSL = Operation.TDSL
module NTDSL = Operation.NTDSL
module CDSL = Train.CDSL
module Asgns = Ir.Assignments
module Tn = Ir.Tnode

module type Backend = Ir.Backend_intf.Backend

let () =
  (* Very similar to micrograd_demo.ml, but with an einsum shape inference corner case. *)
  Tensor.unsafe_reinitialize ();
  let module Backend = (val Backends.fresh_backend ()) in
  let open Operation.At in
  let len = 200 in
  let batch_size = 10 in
  let n_batches = 2 * len / batch_size in
  let moons_config = Datasets.Half_moons.Config.{ noise_range = 0.1; seed = Some 5 } in
  let moons_coordinates, moons_labels =
    Datasets.Half_moons.generate_single_prec ~config:moons_config ~len ()
  in
  let moons_flat_ndarray = Ir.Ndarray.as_array Ir.Ops.Single moons_coordinates in
  let moons_classes_ndarray = Ir.Ndarray.as_array Ir.Ops.Single moons_labels in
  let batch_n, bindings = IDX.get_static_symbol ~static_range:n_batches IDX.empty in
  let step_n, bindings = IDX.get_static_symbol bindings in
  let moons_flat = TDSL.rebatch ~l:"moons_flat" moons_flat_ndarray () in
  let moons_classes = TDSL.rebatch ~l:"moons_classes" moons_classes_ndarray () in
  (* let%op mlp x = 0.5 + ("w3" * relu ("b2" 16 + ("w2" * relu ("b1" 16 + ("w1" * x))))) in *)
  let%op mlp x = "w2" * relu ("b1" 16 + ("w1" * x)) in
  (* Don't decay the learning rate too quickly, it behaves better than in the original. *)
  let%op moons_input = moons_flat @| batch_n in
  (* THIS IS THE SPECIFIC SHAPE INFERENCE ASPECT OF THE TEST. *)
  let%cd _ = moons_input =: 0 ++ "i=>2|i" in
  let%op moons_class = moons_classes @| batch_n in
  let%cd _ = moons_class =: 0 ++ "i=>2|i" in
  let%op margin_loss = relu (1 - (moons_class *. mlp moons_input)) in
  (* We don't need a regression loss formula thanks to weight_decay built into the sgd_update
     computation. *)
  let weight_decay = 0.0001 in
  let%op scalar_loss = (margin_loss ++ "...|... => 0") /. !..batch_size in
  let update = Train.grad_update scalar_loss in
  let%op learning_rate = 0.1 *. ((2 *. !..len) - !@step_n) /. !..len in
  Train.set_hosted learning_rate.value;
  let sgd = Train.sgd_update ~learning_rate ~weight_decay scalar_loss in
  let ctx = Train.init_params (module Backend) bindings scalar_loss in
  let sgd_routine =
    Train.to_routine (module Backend) ctx bindings (Asgns.sequence [ update; sgd ])
  in
  (* Skipping over the training loop, not needed for the test. *)
  let points = Tn.points_2d ~xdim:0 ~ydim:1 moons_flat.value in
  let classes = Tn.points_1d ~xdim:0 moons_classes.value in
  let points1, points2 = Array.partitioni_tf points ~f:Float.(fun i _ -> classes.(i) > 0.) in
  let%cd mlp_result = mlp "point" in
  let result_routine =
    Train.to_routine
      (module Backend)
      sgd_routine.context IDX.empty
      [%cd
        ~~("moons infer";
           mlp_result.forward)]
  in
  let callback (x, y) =
    Tn.set_values point.value [| x; y |];
    Train.run result_routine;
    Float.(mlp_result.@[0] >= 0.)
  in
  let _plot_moons =
    PrintBox_utils.plot ~as_canvas:true ~size:(5, 5)
      [
        Scatterplot { points = points1; content = PrintBox.line "#" };
        Scatterplot { points = points2; content = PrintBox.line "%" };
        Boundary_map
          { content_false = PrintBox.line "."; content_true = PrintBox.line "*"; callback };
      ]
  in
  (* PrintBox_text.output Stdio.stdout _plot_moons; *)
  (* Stdio.printf "Losses: %.4g, %.4g\n%!" losses.(epochs / 2) losses.(epochs - 1); *)
  Tn.print_accessible_headers ();

  (* Testing how the syntax extension %op creates labels for the resulting tensors: *)
  Stdio.printf "mlp_result's name: %s\n%!" @@ Tensor.debug_name mlp_result;
  (* Note: mlp_result is not included in the resulting tensor's label, because the identifier label
     does not propagate across function calls. *)
  Stdio.printf "(mlp moons_input) name: %s\n%!"
  @@ Tensor.debug_name
  @@
  match margin_loss.children with
  | [
   {
     subtensor =
       { children = [ _; { subtensor = { children = [ _; { subtensor; _ } ]; _ }; _ } ]; _ };
     _;
   };
  ] ->
      subtensor
  | _ -> assert false
