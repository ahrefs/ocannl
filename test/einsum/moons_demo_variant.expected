
Welcome to OCANNL! Reading configuration defaults from /Users/lukstafi/ocannl/_build/default/test/einsum/ocannl_config.
Retrieving commandline, environment, or config file variable ocannl_log_level
Found 0, in the config file
Tnode: collecting accessible arrays...
n5 grad_b2 as b2.grad: Local/26046; single prec 16; mem in bytes: <not-in-yet>
n22 + as n22: Local/1046; double prec 10x16; mem in bytes: <not-in-yet>
n32 * as n32: Local/1046; double prec 10x1; mem in bytes: <not-in-yet>
n34 +_mlp_@|_moons_input as mlp_moons_input: Virt/15; double prec 10x1; mem in bytes: <not-in-yet>
n39 - as n39: Local/1046; double prec 10x1; mem in bytes: <not-in-yet>
n51 *. as n51: Virt/152; double prec 1; mem in bytes: <not-in-yet>
n59 - as n59: Virt/152; single prec 1; mem in bytes: <not-in-yet>
n64 sgd_momentum_b1 as sgd_momentum_b1: unknown; single prec <not-in-yet>; mem in bytes: <not-in-yet>
n70 *. as n70: Virt/15; single prec 16; mem in bytes: <not-in-yet>
n78 *. as n78: Virt/15; single prec 16x2; mem in bytes: <not-in-yet>
n83 sgd_delta_w3 as sgd_delta_w3: Virt/15; single prec 1x16; mem in bytes: <not-in-yet>
n84 sgd_momentum_w3 as sgd_momentum_w3: unknown; single prec <not-in-yet>; mem in bytes: <not-in-yet>
n93 grad_relu as relu.grad: unknown; single prec 16; mem in bytes: <not-in-yet>
n4 b2 as b2: Host&stream/412410; single prec 16; mem in bytes: 64
n13 grad_w3 as w3.grad: Local/26046; single prec 1x16; mem in bytes: <not-in-yet>
n47 grad_/._scalar_loss as scalar_loss.grad: Virt/40; single prec 1; mem in bytes: <not-in-yet>
n49 **. as n49: Virt/40; single prec 1; mem in bytes: <not-in-yet>
n50 -1. as n50: Virt/40; single prec 1; mem in bytes: <not-in-yet>
n56 80. as n56: Virt/40; single prec 1; mem in bytes: <not-in-yet>
n61 *. as n61: Virt/152; single prec 1; mem in bytes: <not-in-yet>
n79 sgd_delta_w2 as sgd_delta_w2: Virt/15; single prec 16x16; mem in bytes: <not-in-yet>
n92 relu as relu: Local/1046; single prec 16; mem in bytes: <not-in-yet>
n6 b3 as b3: Host&stream/412410; single prec 1; mem in bytes: 4
n55 !@ as n55: Virt/152; single prec 1; mem in bytes: <not-in-yet>
n65 0.0001 as n65: Virt/40; single prec 1; mem in bytes: <not-in-yet>
n29 grad_+ as n28.grad: Local/1046; single prec 10x16; mem in bytes: <not-in-yet>
n35 grad_+_mlp_@|_moons_input as mlp_moons_input.grad: Local/1046; single prec 10x1; mem in bytes: <not-in-yet>
n41 relu_margin_loss as relu_margin_loss: Virt/15; double prec 10x1; mem in bytes: <not-in-yet>
n77 0.0001 as n77: Virt/40; single prec 1; mem in bytes: <not-in-yet>
n94 * as n94: Local/1046; single prec 16; mem in bytes: <not-in-yet>
n7 grad_b3 as b3.grad: Local/26046; single prec 1; mem in bytes: <not-in-yet>
n20 * as n20: Local/1046; double prec 10x16; mem in bytes: <not-in-yet>
n21 grad_* as n20.grad: Local/1046; single prec 10x16; mem in bytes: <not-in-yet>
n33 grad_* as n32.grad: Local/1046; single prec 10x1; mem in bytes: <not-in-yet>
n43 10. as n43: Virt/40; single prec 1; mem in bytes: <not-in-yet>
n54 80. as n54: Virt/40; single prec 1; mem in bytes: <not-in-yet>
n63 sgd_delta_b1 as sgd_delta_b1: Virt/15; single prec 16; mem in bytes: <not-in-yet>
n67 sgd_delta_b2 as sgd_delta_b2: Virt/15; single prec 16; mem in bytes: <not-in-yet>
n81 0.0001 as n81: Virt/40; single prec 1; mem in bytes: <not-in-yet>
n86 *. as n86: Virt/15; single prec 1x16; mem in bytes: <not-in-yet>
n95 grad_* as n94.grad: unknown; single prec 16; mem in bytes: <not-in-yet>
n2 b1 as b1: Host&stream/412410; single prec 16; mem in bytes: 64
n11 grad_w2 as w2.grad: Local/26046; single prec 16x16; mem in bytes: <not-in-yet>
n25 grad_relu as relu.grad: Local/1046; single prec 10x16; mem in bytes: <not-in-yet>
n45 grad_=> as n44.grad: Virt/40; single prec 1; mem in bytes: <not-in-yet>
n69 0.0001 as n69: Virt/40; single prec 1; mem in bytes: <not-in-yet>
n10 w2 as w2: Host&stream/412410; single prec 16x16; mem in bytes: 1_024
n27 grad_* as n26.grad: Local/1046; single prec 10x16; mem in bytes: <not-in-yet>
n36 *. as n36: Virt/15; double prec 10x1; mem in bytes: <not-in-yet>
n44 => as n44: Local/1046; double prec 1; mem in bytes: <not-in-yet>
n58 *. as n58: Virt/40; single prec 1; mem in bytes: <not-in-yet>
n68 sgd_momentum_b2 as sgd_momentum_b2: unknown; single prec <not-in-yet>; mem in bytes: <not-in-yet>
n97 grad_+ as n96.grad: unknown; single prec 16; mem in bytes: <not-in-yet>
n52 /. as n52: Host&stream/412410; double prec 1; mem in bytes: <not-in-yet>
n38 1. as n38: Virt/40; single prec 1; mem in bytes: <not-in-yet>
n88 * as n88: Local/1046; single prec 16; mem in bytes: <not-in-yet>
n99 grad_relu as relu.grad: unknown; single prec 16; mem in bytes: <not-in-yet>
n9 grad_w1 as w1.grad: Local/26046; single prec 16x2; mem in bytes: <not-in-yet>
n26 * as n26: Local/1046; double prec 10x16; mem in bytes: <not-in-yet>
n31 grad_relu as relu.grad: Local/1046; single prec 10x16; mem in bytes: <not-in-yet>
n60 0.1 as n60: Virt/40; single prec 1; mem in bytes: <not-in-yet>
n72 sgd_momentum_b3 as sgd_momentum_b3: unknown; single prec <not-in-yet>; mem in bytes: <not-in-yet>
n98 relu as relu: Virt/15; single prec 16; mem in bytes: <not-in-yet>
n0 moons_flat as moons_flat: Host-const/37; double prec 40x10x2; mem in bytes: 6_400
n12 w3 as w3: Host&stream/412410; single prec 1x16; mem in bytes: 64
n46 /._scalar_loss as scalar_loss: Host&stream/412410; double prec 1; mem in bytes: 8
n71 sgd_delta_b3 as sgd_delta_b3: Virt/152; single prec 1; mem in bytes: <not-in-yet>
n73 0.0001 as n73: Virt/40; single prec 1; mem in bytes: <not-in-yet>
n90 + as n90: Virt/15; single prec 16; mem in bytes: <not-in-yet>
n24 relu as relu: Local/1046; double prec 10x16; mem in bytes: <not-in-yet>
n30 relu as relu: Local/1046; double prec 10x16; mem in bytes: <not-in-yet>
n82 *. as n82: Virt/15; single prec 16x16; mem in bytes: <not-in-yet>
n85 0.0001 as n85: Virt/40; single prec 1; mem in bytes: <not-in-yet>
n91 grad_+ as n90.grad: unknown; single prec 16; mem in bytes: <not-in-yet>
n28 + as n28: Local/1046; double prec 10x16; mem in bytes: <not-in-yet>
n42 grad_relu_margin_loss as relu_margin_loss.grad: Local/1046; single prec 10x1; mem in bytes: <not-in-yet>
n62 /._learning_rate as learning_rate: Host&stream/412410; single prec 1; mem in bytes: <not-in-yet>
n66 *. as n66: Virt/15; single prec 16; mem in bytes: <not-in-yet>
n87 point as point: Host-const/37; single prec 2; mem in bytes: 8
n89 grad_* as n88.grad: unknown; single prec 16; mem in bytes: <not-in-yet>
n8 w1 as w1: Host&stream/412410; single prec 16x2; mem in bytes: 128
n14 @|_moons_input as moons_input: Local/1046; double prec 10x2; mem in bytes: <not-in-yet>
n74 *. as n74: Virt/152; single prec 1; mem in bytes: <not-in-yet>
n75 sgd_delta_w1 as sgd_delta_w1: Virt/15; single prec 16x2; mem in bytes: <not-in-yet>
n17 @|_moons_class as moons_class: Local/1046; double prec 10x1; mem in bytes: <not-in-yet>
n76 sgd_momentum_w1 as sgd_momentum_w1: unknown; single prec <not-in-yet>; mem in bytes: <not-in-yet>
n103 grad_+_mlp_point as mlp_point.grad: unknown; single prec 1; mem in bytes: <not-in-yet>
n3 grad_b1 as b1.grad: Local/26046; single prec 16; mem in bytes: <not-in-yet>
n37 grad_*. as n36.grad: Local/1046; single prec 10x1; mem in bytes: <not-in-yet>
n57 2. as n57: Virt/40; single prec 1; mem in bytes: <not-in-yet>
n80 sgd_momentum_w2 as sgd_momentum_w2: unknown; single prec <not-in-yet>; mem in bytes: <not-in-yet>
n102 +_mlp_point as mlp_point: Host&stream/412410; single prec 1; mem in bytes: <not-in-yet>
n1 moons_classes as moons_classes: Host-const/37; double prec 40x10x1; mem in bytes: 3_200
n23 grad_+ as n22.grad: Local/1046; single prec 10x16; mem in bytes: <not-in-yet>
n40 grad_- as n39.grad: Local/1046; single prec 10x1; mem in bytes: <not-in-yet>
n100 * as n100: Local/1046; single prec 1; mem in bytes: <not-in-yet>
n48 2. as n48: Virt/40; single prec 1; mem in bytes: <not-in-yet>
n53 1. as n53: Virt/40; single prec 1; mem in bytes: <not-in-yet>
n96 + as n96: Virt/15; single prec 16; mem in bytes: <not-in-yet>
n101 grad_* as n100.grad: unknown; single prec 1; mem in bytes: <not-in-yet>
Tnode: Finished printing headers.
mlp_result's name: mlp_point
(mlp moons_input) name: mlp_moons_input
